pkg/server/server_podman_test.go

```diff
@@ -0,0 +1,343 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/containers/libpod/v2/libpod/define"
+	"github.com/containers/libpod/v2/pkg/bindings/containers"
+	"github.com/containers/libpod/v2/pkg/bindings/pods"
+	"github.com/elotl/itzo/pkg/api"
+	"github.com/elotl/itzo/pkg/runtime"
+	"github.com/elotl/wsstream"
+	"github.com/stretchr/testify/assert"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"strings"
+	"testing"
+	"time"
+)
+// Here are E2E test for server endpoints and podman runtime.
+// Before running those you need to:
+// 1. Ensure that podman.socket is open for superuser by running sudo systemctl start podman.socket
+// 2. Ensure that sudo podman pod ps doesn't return itzpod - that's a constant that we use for pod created by itzo.
+// 3. Executing of those test may take longer as they're E2E; using podman API we create, run, stop and remove pods and containers here.
+//    All podman resources should be removed after each test with removeContainersAndPods function.
+
+var (
+	stateStopped = define.ContainerStateStopped
+	stateRunning = define.ContainerStateRunning
+	podmanAPITimeout uint = 5
+	forceContainerRemove = true
+	removeVolumes = true
+)
+
+func setUpServerAndController() (*Server, *PodController, error) {
+	tmpdir, err := ioutil.TempDir("", "itzo-test-podman")
+	if err != nil {
+		panic("Error creating temporary directory")
+	}
+	defer os.RemoveAll(tmpdir)
+	podctl, err := NewPodController(tmpdir, true)
+	testServer := NewTestServer(EnvStore{}, tmpdir, podctl)
+	podctl.Start()
+	return &testServer, podctl, err
+}
+
+func tearDownSeverAndController() {
+	KillChildren()
+	return
+}
+
+func removeContainersAndPod(containersNames []string) {
+	conn, err := runtime.GetPodmanConnection()
+	if err != nil {
+		panic(err)
+	}
+	for _, name := range containersNames {
+		if !strings.HasPrefix(name, api.PodName) {
+			name = api.PodName + "-" + name
+		}
+		err = containers.Stop(conn, name, &podmanAPITimeout)
+		if err != nil {
+			panic(err)
+		}
+		_, err = containers.Wait(conn, name, &stateStopped)
+		if err != nil {
+			panic(err)
+		}
+		err = containers.Remove(conn, name, &forceContainerRemove, &removeVolumes)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	timeout := int(podmanAPITimeout)
+	_, err = pods.Stop(conn, api.PodName, &timeout)
+	if err != nil {
+		panic(err)
+	}
+	_, err = pods.Remove(conn, api.PodName, &forceContainerRemove)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
Shouldn't we be able to remove containers and the pod via sending a pod update with no containers?